### ISSUE TITLE
Rex on Mac invokes wrong MD5 program

### ISSUE THREAD

**Issue by einhverfr** [RexOps/Rex#1187]
see https://github.com/Homebrew/homebrew-core/pull/22564

Rex is invoking md5sum when md5 is what is included.

**Comment by ferki** [comment]
Hi @einhverfr, thanks for the report! I think I would need more context to understand the exact goal here and in the homebrew repo.

**TL;DR version:**

Since I don't have access to Mac OS currently, could you post the following info from Mac, please?

 - output of `uname` (or, at least, `uname -s`)
 - path to `md5` (output of `which md5`)
 - example output of `md5 $filename` (to see how it can be parsed)

**Longer version:**

Currently Rex only needs to calculate MD5 checksums of files on the target machine via the `Digest::MD5` Perl module (if available), or either `md5` (on BSDs) or `md5sum` (on non-BSDs) CLI utilities (in other words, it's an "any-of" type of remote dependency).

AFAIK, managing Mac machines with Rex is not (yet :) possible/supported, so if that's the goal, probably calculating MD5 checksums would be just the tip of the iceberg (and for example, package and service management needs to be implemented too).

That being said, running Rex on Mac to manage _other_ remote targets should work, so if you are seeing, for example, test failures while installing, then that might lead us to a slightly different fix.

**Comment by vinyldarkscratch** [comment]
Hey @ferki,

Going through [homebrew-core](https://github.com/Homebrew/homebrew-core) issues, I noticed this hadn't been touched in a while.  To help out with this:

uname [-s]: `Darwin`
MD5 Path: `/sbin/md5`
Example Output: `MD5 (/Users/vinyldarkscratch/Desktop/whs.glitch.extract.js) = 6a6c199984785367ed666504846442cd`

**Comment by ferki** [comment]
Thanks for the patience on this one, it turned out there's a bunch of sides to this story.

First, the [Homebrew formula for Rex](https://github.com/Homebrew/homebrew-core/blob/bc253ed78337eaa551bc5abfb4a3a4753b3c5c8b/Formula/rex.rb) is (still) missing required dependencies, including the Digest::MD5 module that is relevant for this issue. This should be fixed on the packaging side, so it's a good opportunity for someone with Mac OS X or Homebrew experience to get involved.

Second, if obtaining the MD5 checksum through the Digest::MD5 module fails, Rex tries to use a binary as a fallback option. Currently this depends on the OS: it tries `md5` on BSDs, and `md5sum` anywhere else. With that in mind, a possible hotfix for Rex::Commands::MD5 could be the following (untested though):

```diff
-      if ( $os =~ /bsd/i ) {
+      if ( $os =~ /bsd|darwin/i ) {
```

Third, in order to make the (fallback) behavior testable, I decided to refactor Rex::Commands::MD5, and after adding a couple of further tests, I modified the logic to choose a binary based on availability instead of the operating system. On Travis CI, Mac OS X has both `md5` and `md5sum` available, and even Windows has an `md5sum.exe` (with a slightly different output format though).

I'll open a PR soon for this. Until then, let me know your experience with the hotfix if you try that.
